DigitalExcellence/dex-backend
# Feature/287 user graduation detection

## Description

This branch has:
- added a new controller called UserTaskController which is responsible for user tasks.
In our case a user task is converting a user account to alumni role which needs user input to complete.

One of the endpoint is ConvertToAlumni. This endpoint checks if the user meets the requirements. When it does it makes a call to the identityserver to change login and password. If this is succesfull the account is converted to Alumni.

- added a new project called JobScheduler which is responsible for handling scheduled jobs.
The project contains a job which does a Rest request to the API. It asks for all graduating users that do not yet have a task for it.
During the call the API creates the tasks. When the notification system is ready it will be used to queue emails to expected graduating users.

## Notes

ProjectWrite scope was functionality is now changed: ProjectWrite scope is required to create projects.
Before the functionality was to edit projects that are not yours. A new scope was created for that called AdminProjectWrite.
In the seed this is fixed, but in staging and production we should manually add the scopes to the correct roles:
Add ProjectWrite scope to all roles except Alumni.
Add AdminProjectWrite scope to role Administrator.

## Type of change


## Checklist


## Steps to Test or Reproduce

1. Run code
2. In the user table make sure there is at least one user that has ExpectedGraduationDate within now and next 6 months.
3. Make a get call to /api/UserTask
4. Receive user mentioned in step 2. 
5. Make a put call to /api/UserTask with the new credentials in the header

## Link to issue

Closes: #287
